### ISSUE TITLE
[lx] mmap is executable by default, even without READ_IMPLIES_EXEC personality bit set

### ISSUE THREAD

**Issue by nathan7** [joyent/illumos-joyent#99]
# [the culprit](https://github.com/joyent/illumos-joyent/blob/868d62fdba8df95378d6e17a5b6154e3fd228de9/usr/src/lib/brand/lx/lx_brand/common/mem.c#L134-L144)

# Test case
```c
#include <assert.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/personality.h>

unsigned char shellcode[] = {
  0xbf, 0x01, 0x00, 0x00, 0x00, 0x48, 0x8d, 0x35, 0x0d, 0x00, 0x00, 0x00,
  0xba, 0x19, 0x00, 0x00, 0x00, 0xb8, 0x01, 0x00, 0x00, 0x00, 0x0f, 0x05,
  0xc3, 0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x2c, 0x20, 0x65, 0x78, 0x65, 0x63,
  0x75, 0x74, 0x61, 0x62, 0x6c, 0x65, 0x20, 0x77, 0x6f, 0x72, 0x6c, 0x64,
  0x21, 0x0a
};

int main(int argc, char **argv) {
  int personality_flags = personality(-1);
#ifdef COMPAT
  personality_flags |= READ_IMPLIES_EXEC;
#else
  personality_flags &= ~READ_IMPLIES_EXEC;
#endif
  assert(personality(personality_flags) != -1);

  void* mapping = mmap((void*) 0, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0);
  assert(mapping != MAP_FAILED);
  memcpy(mapping, shellcode, sizeof shellcode);
  ((void(*)())mapping)();
  return 0;
}
```

## Desired behaviour (observed on native Linux)
```
[nix-shell:~]$ uname -a
Linux koala 4.3.6 #1-NixOS SMP Fri Feb 19 22:35:24 UTC 2016 x86_64 GNU/Linux
[nix-shell:~]$ cc test.c -o test -DCOMPAT
[nix-shell:~]$ ./test 
Hello, executable world!
[nix-shell:~]$ cc test.c -o test
[nix-shell:~]$ ./test 
Segmentation fault
```

## Observed behaviour on SmartOS
```
[nix-shell:~]$ uname -a
Linux 991490c2-bf90-479b-dfeb-fd9a723a9ba4 3.13.0 BrandZ virtual linux x86_64 GNU/Linux
[nix-shell:~]$ cc test.c -o test -DCOMPAT
[nix-shell:~]$ ./test 
Hello, executable world!
[nix-shell:~]$ cc test.c -o test
[nix-shell:~]$ ./test 
Hello, executable world!
```

**Comment by nathan7** [comment]
The equivalent in Linux uses `unlikely(personality & READ_IMPLIES_EXEC)` to inform the compiler that it's a cold codepath, but I'm not sure how to do the same thing in illumos code.

**Comment by pfmooney** [comment]
I'll get this filed and take a closer look in the morning (UTC-5).

**Comment by pfmooney** [comment]
This should be fixed in fcc3d074e6a2ad9529ce014e16d99720f33a053d

Thanks for the detailed report.
